### PR TITLE
[AssetSubset] Make ValidAssetSubset a real class

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Previously, this was just a hint for the type system. This makes it a real subclass of the AssetSubset class.

You can no longer use any of the overloaded operators on the base AssetSubset class (because we can't be sure that their underlying values will be identical).

You _are_ able to do ValidAssetSubset | AssetSubset, in which case the ValidAssetSubset will validate if the other subset is compatible. If it is, it just does the normal operation, and if it's not, it converts the other subset into the empty subset and goes from there.

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/asset_subset.py [hunk 1]
```diff
@@ -131,7 +151,17 @@ def from_asset_partitions_set(
...
+
+
+@whitelist_for_serdes(serializer=AssetSubsetSerializer, storage_name="AssetSubset")
+class ValidAssetSubset(AssetSubset):
+    """Represents an AssetSubset which is known to be compatible with the current
+    PartitionsDefinition of the asset represents.
+
+    This class serializes to a regular AssetSubset, because once it is deserialized we can't be
+    certain that the subset will still be valid.
+    """
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/asset_subset.py:164]: Can you add a little test that just runs through the expectations for serialization / deserialization ? This setup is a little funky so want to protect it from future changes.

Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_subset.py]: good callout -- realized that with this storage_name thing, all AssetSubsets were getting deserialized as ValidAssetSubsets due to weirdness with the WhitelistMap not allowing duplicate storage names. Fixed that and updated tests

File: python_modules/dagster/dagster/_core/definitions/asset_subset.py [hunk 2]
```diff
@@ -150,56 +187,32 @@ def inverse(
...
+        valid_other = self.get_valid(other)
         if not self.is_partitioned:
-            return other._replace(value=self.bool_value and not other.bool_value)
-        return self._oper(other, operator.sub)
+            return self._replace(value=self.bool_value and not valid_other.bool_value)
+        return self._oper(valid_other, operator.sub)
 
-    def __and__(self, other: "ValidAssetSubset") -> "ValidAssetSubset":
+    def __and__(self, other: AssetSubset) -> "ValidAssetSubset":
         """Returns the intersection of this AssetSubset and another AssetSubset if they are compatible,
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_subset.py:206]: Is there a way to rewrite these comment on these operators in a way that makes it clear how they can be interpreted as acting on sets?

**Comment by OwenKephart** [comment]
https://github.com/dagster-io/dagster/pull/19483 original PR, for context

File: python_modules/dagster/dagster/_core/definitions/asset_subset.py (hunk 1 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/_core/definitions/asset_subset.py]: do you think  the serdes whitelist map thing should throw an error based on your journey here?

File: python_modules/dagster/dagster/_core/definitions/asset_subset.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_subset.py]: @sryza updated!

File: python_modules/dagster/dagster/_core/definitions/asset_subset.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_subset.py]: Hm good question, certainly doesn't hurt: https://github.com/dagster-io/dagster/pull/19651/files

**Review by sryza — APPROVED** [review]

**Pull request merged**